zschema/compounds.py

```diff
@@ -10,9 +10,10 @@ def _is_valid_object(name, object_):
 
 class ListOf(Keyable):
 
-    def __init__(self, object_, max_items=10):
+    def __init__(self, object_, max_items=10, category=None):
```

Comment:
It looks like `ListOf` doesn't recognize `doc`; should that be pushed up to `Keyable`?

Reply:
Another example -- the SSH logs include both the client `EndpointId` and the server `EndpointId`, which are the same struct at the go level, but in the schema, one contains `AnalyzedString()` types, while the other has `String()` types.

Reply:
> It looks like `ListOf` doesn't recognize `doc`; should that be pushed up to `Keyable`?

My initial thinking here was that a list just lists things of a given type, so the documentation from the type being listed ought to be sufficient. Thinking more, that is wrong: a list of strings has a distinct meaning from a string. It makes sense for a list itself to have documentation describing what the list actually does — the context for the objects of this type — and that should be easy to add (though even if some part of this moves up to `Keyable`, we'll still have to modify the initializer for `ListOf`).

> So maybe the problem is using `SubRecord()`, which returns an instance of the type instead of returning a constructor that can be used like `String()` ; i.e. we need something like `Certificate = SubRecordType({ ... definition ...})`, then `"IssuerCert": Certificate(doc="The server certificate's issuer")`. That might fit better with the current model than trying to introduce separate docs for fields and types.
>
> …
> 
> This actually dovetails with a broader issue I've come across, where we are attaching things like `doc` to the *types*, where they should really be attached to *fields*.

If I can rephrase, to be sure we're on the same page: a record definition that gets reused, like `Certificate`, currently cannot have different documentation depending on the context where it is reused.

This is a definite limitation that it's worth calling out. My thinking here is basically "hopefully it's clear from context what the (certificate, etc.) is," eg. the subject and issuer certs can be differentiated by the fact that one is named "subject" and the other is named "issuer."

Removing this limitation would be a major change that would, I think, require some nontrivial changes to zschema and major changes to our schema. I am not sure we want to undertake that right now.

Reply:
> Removing this limitation would be a major change that would, I think, require some nontrivial changes to zschema and major changes to our schema. I am not sure we want to undertake that right now.

I agree with that, @cdzombak; nice to have, and _should_ happen, but out of scope of this PR.

>  It makes sense for a list itself to have documentation describing what the list actually does — the context for the objects of this type — and that should be easy to add (though even if some part of this moves up to `Keyable`, we'll still have to modify the initializer for `ListOf`).

@justinbastress & @cdzombak: do we want to implement that as part of this PR?

Reply:
> do we want to implement that as part of this PR?

I'll work on it now.

Reply:
https://github.com/zmap/zschema/pull/18/commits/5be9e2634c6b08fb62ba13dbc8e97ed2c4c06baa introduces a `doc` field for `ListOf`.